### PR TITLE
Splits reagent metabolization from reagent effects.

### LINKED ISSUE #79267: Growth serum does not have a cap on increasing mob size and will look extremely funky and broken when vaped to the extreme
Reporting client version: 514.1589

## Round ID:
[217479](https://scrubby.melonmesa.com/round/217479)

## Testmerges:
- [Makes lighting objects objects again](https://www.github.com/tgstation/tgstation/pull/78857)
- [Cuts the number of `apply_damage` copypaste procs from 3(.5) to 1, fixing a few bugs along the way](https://www.github.com/tgstation/tgstation/pull/79207)

## Reproduction:
1. acquire a vape with growth serum (I do not actually know how this was accomplished on this particular round, but I believe it may have been botany growing Fly Amanita plants? regardless, they were handing out vapes with growth serum in them)
2. put it in your mouth and suck
3. achieve a puberty level growth spurt every 5-10 seconds, using up miniscule quantities of the reagent each time
4. realize there is no upper bound on this growth

Me after using the vape for a few minutes on view range 10, using up less than 5 units of growth serum

### PR DESCRIPTION
**Pull request by Ghommie**
## About The Pull Request
I've separated the instructions responsible for gradually removing reagents from a living holder from the main `on_mob_life` and `on_mob_dead` procs, and made sure they're run after them.

I've done a fair bit of early returns as well.

## Why It's Good For The Game
The problem here is that most reagents will eventually be deleted before  `on_mob_life` and `on_mob_dead` have finished running, since the statements that handle the reagent removal are found within parent calls that are often called at the start of the different definitions of the proc . To make it easier to understand, here's a simplified chain of events:

holder.metabolize_reagent -> the parent call of reagent.on_mob_life -> holder.remove_reagent -> holder.update_total -> reagent.on_mob_end_metabolizate -> reagent.on_mob_delete -> qdel(reagent) -> the actual reagent.on_mob_life

Basically, the reagent is getting deleted before it has finished "ticking", and that's bad, simple as that.

Also, by separating these instructions from the rest, we won't have to enforce any such nuisance as `return ..() || need_mob_update`.

TL;DR: This fixes #79267.

## Changelog

:cl:
fix: Fixed the infinite growth serum exploit.
fix: Fixed generic nutriment processing even when dead.
/:cl:

### CODE REVIEW

File: code/modules/reagents/chemistry/reagents.dm [hunk 1]
```diff
@@ -154,8 +154,10 @@
  *
  */
 /datum/reagent/proc/on_mob_life(mob/living/carbon/affected_mob, seconds_per_tick, times_fired)
-	SHOULD_CALL_PARENT(TRUE)
```
Comment (JohnFulpWillard) [inline code/modules/reagents/chemistry/reagents.dm:157]: Even removing the code, I think it's still a good idea to have ``SHOULD_CALL_PARENT`` in case something is added in the future.

File: code/modules/reagents/chemistry/reagents/other_reagents.dm [hunk 2]
```diff
@@ -2458,6 +2434,8 @@
 	affected_mob.update_transform(RESIZE_DEFAULT_SIZE/current_size)
 	current_size = RESIZE_DEFAULT_SIZE
 
+/datum/reagent/growthserum/on_mob_delete(mob/living/affected_mob)
```
Comment (JohnFulpWillard) [inline code/modules/reagents/chemistry/reagents/other_reagents.dm:2437]: What's this here for?

File: code/modules/reagents/chemistry/reagents/pyrotechnic_reagents.dm [hunk 3]
```diff
@@ -333,16 +325,15 @@
...
-	if(isjellyperson(affected_mob))
-		shock_timer = 0 //immune to shocks
-		affected_mob.AdjustAllImmobility(-40  *REM * seconds_per_tick)
-		if(affected_mob.adjustStaminaLoss(-2 * REM * seconds_per_tick, updating_stamina = FALSE))
-			. = UPDATE_MOB_HEALTH
-		if(is_species(affected_mob, /datum/species/jelly/luminescent))
-			var/mob/living/carbon/human/affected_human = affected_mob
-			var/datum/species/jelly/luminescent/slime_species = affected_human.dna.species
-			slime_species.extract_cooldown = max(slime_species.extract_cooldown - (2 SECONDS * REM * seconds_per_tick), 0)
+	if(!isjellyperson(affected_mob)) //jellypeople don't get shocked.
```
Comment (JohnFulpWillard) [inline code/modules/reagents/chemistry/reagents/pyrotechnic_reagents.dm:328]: This comment makes it seem like it's early returning to prevent jellypeople from getting shocked but it's kinda inaccurate, maybe this is better

```suggestion
	if(!isjellyperson(affected_mob)) //everyone but jellypeople get shocked as normal.
```

**Comment by san7890** [comment]
conflicts
